compiler-cli/src/dependencies.rs

```diff
@@ -678,6 +633,237 @@ fn same_requirements(
+    // manifest which will result in the latest versions of the dependency
+    // packages being resolved (not the locked ones).
+    pub use_manifest: UseManifest,
+    /// When set to `Yes`, the cli will check for major version updates of direct dependencies and
+    /// print them to the console if the major versions are not upgradeable due to constraints.
+    pub check_major_versions: CheckMajorVersions,
+}
+
+impl DependencyManagerConfig {
+    fn into_dependency_manager<Telem: Telemetry, P: dependency::PackageFetcher>(
+        self,
+        runtime: tokio::runtime::Handle,
+        package_fetcher: P,
+        telemetry: Telem,
+        mode: Mode,
+    ) -> DependencyManager<Telem, P> {
+        DependencyManager {
+            runtime,
+            package_fetcher,
+            telemetry,
+
+            mode,
+            use_manifest: self.use_manifest,
+            check_major_versions: self.check_major_versions,
+        }
+    }
+}
+
+struct DependencyManager<Telem, P> {
+    runtime: tokio::runtime::Handle,
+    package_fetcher: P,
+    mode: Mode,
+    use_manifest: UseManifest,
+    telemetry: Telem,
+    check_major_versions: CheckMajorVersions,
+}
+
+impl<Telem, P> DependencyManager<Telem, P>
+where
+    P: dependency::PackageFetcher,
+    Telem: Telemetry,
+{
+    fn get_manifest(
+        &self,
+        paths: &ProjectPaths,
+        config: &PackageConfig,
+        packages_to_update: Vec<EcoString>,
+    ) -> Result<(bool, Manifest)> {
+        // If there's no manifest (or we have been asked not to use it) then resolve
+        // the versions anew
+        let should_resolve = match self.use_manifest {
+            _ if !paths.manifest().exists() => {
+                tracing::debug!("manifest_not_present");
+                true
+            }
+            UseManifest::No => {
+                tracing::debug!("ignoring_manifest");
+                true
+            }
+            UseManifest::Yes => false,
+        };
+
+        if should_resolve {
+            let manifest = self.resolve_versions(paths, config, None, Vec::new())?;
+            return Ok((true, manifest));
+        }
+
+        let manifest = read_manifest_from_disc(paths)?;
+
+        // If there are no requested updates, and the config is unchanged
+        // since the manifest was written then it is up to date so we can return it unmodified.
+        if packages_to_update.is_empty()
+            && is_same_requirements(
+                &manifest.requirements,
+                &config.all_direct_dependencies()?,
+                paths.root(),
+            )?
+        {
+            tracing::debug!("manifest_up_to_date");
+            Ok((false, manifest))
+        } else {
+            tracing::debug!("manifest_outdated");
+            let manifest =
+                self.resolve_versions(paths, config, Some(&manifest), packages_to_update)?;
+            Ok((true, manifest))
+        }
+    }
+
+    fn download(
```

Comment:
It was tricky for me to determine that this was the public API of this class. Could you mark this as public and move the class to sub-module please 🙏 

I'm concerned about the API being understandable, particularly since the locking has been moved inside this class, so it is extremely important that the correct entrypoint is used.

Reply:
Done.